pkg/openthread/contrib/sock/udp/openthread_sock_udp.c

```diff
@@ -0,0 +1,387 @@
+    otMessage *message;
+
+    message = otUdpNewMessage(ot_instance, true);
+    int error;
+    error = otMessageSetLength(message, pkt->len);
+    error = otMessageWrite(message, 0, pkt->payload, pkt->len);
+    (void) error;
+
+    otMessageInfo mPeer;
+    
+    //Set dest address
+    memcpy(&mPeer.mPeerAddr.mFields, &(pkt->ip_addr), sizeof(ipv6_addr_t));
+
+    //Set dest port
+    mPeer.mPeerPort = pkt->port;
+
+    otUdpSend(&sock->ot_udp_socket, message, &mPeer);
+}
+
+int sock_udp_create(sock_udp_t *sock, const sock_udp_ep_t *local,
+                    const sock_udp_ep_t *remote, uint16_t flags)
+{
+    uint16_t socket_port;
+
+    assert(sock);
+    assert(local == NULL || local->port != 0);
+    assert(remote == NULL || remote->port != 0);
+    if ((local != NULL) && (remote != NULL) &&
+        (local->netif != SOCK_ADDR_ANY_NETIF) &&
+        (remote->netif != SOCK_ADDR_ANY_NETIF) &&
+        (local->netif != remote->netif)) {
+        return -EINVAL;
+    }
+    
+	mbox_init(&sock->mbox, sock->mbox_queue, SOCK_MBOX_SIZE);
+	sock->data = NULL;
+    memset(&sock->local, 0, sizeof(sock_udp_ep_t));
+    
+    if (local != NULL) {
+        if (openthread_af_not_supported(local->family)) {
+            return -EAFNOSUPPORT;
+        }
+        memcpy(&sock->local, local, sizeof(sock_udp_ep_t));
+    }
+    
+    memset(&sock->remote, 0, sizeof(sock_udp_ep_t));
+    
+    if (remote != NULL) {
+        if (openthread_af_not_supported(remote->family)) {
+            return -EAFNOSUPPORT;
+        }
+        if (openthread_ep_addr_any((const sock_ip_ep_t *)remote)) {
+            return -EINVAL;
+        }
+        memcpy(&sock->remote, remote, sizeof(sock_udp_ep_t));
+    }
+
+    memset(&sock->ot_udp_socket, 0, sizeof(otUdpSocket));
+
+    socket_port = local->port;
+
+    if (local != NULL) {
+        DEBUG("Socket port %d \n", socket_port);
+        ot_exec_job(_create_udp_socket, sock);
+    }
+    else
+    {
+        DEBUG("Null local\n");
+    }
+    sock->flags = flags;
+
+    return 0;
+}
+
+void sock_udp_close(sock_udp_t *sock)
+{
+    assert(sock != NULL);
+    if (sock->ot_udp_socket.mTransport != NULL) {
+        otUdpClose(&sock->ot_udp_socket);
+        memset(&sock->ot_udp_socket, 0, sizeof(otUdpSocket));
+    }
+}
+
+int sock_udp_get_local(sock_udp_t *sock, sock_udp_ep_t *local)
+{
+    assert(sock && local);
+    if (sock->local.family == AF_UNSPEC) {
+        return -EADDRNOTAVAIL;
+    }
+    memcpy(local, &sock->local, sizeof(sock_udp_ep_t));
+    return 0;
+}
+
+int sock_udp_get_remote(sock_udp_t *sock, sock_udp_ep_t *remote)
+{
+    assert(sock && remote);
+    if (sock->remote.family == AF_UNSPEC) {
+        return -ENOTCONN;
+    }
+    memcpy(remote, &sock->remote, sizeof(sock_udp_ep_t));
+    return 0;
+}
+
+ssize_t sock_udp_recv(sock_udp_t *sock, void *data, size_t max_len,
+                      uint32_t timeout, sock_udp_ep_t *remote)
+{
+	// xtimer_t timeout_timer;
```

Comment:
Ok, if this is still WIP, I would suggest to work in smaller steps: most important thing is to have the basic openthread port merged ASAP. Otherwise this will never end.
I propose this:
* Split this PR in 2: one PR with the basic of openthread (as it was before). The idea is at least to use openthread command lines and API on top of RIOT.
* Open a follow-up PR with sock adaption